pkg/operator/controller/certificate-publisher/publish_certs.go

```diff
@@ -56,55 +57,63 @@ func (r *reconciler) ensureRouterCertsGlobalSecret(secrets []corev1.Secret, ingr
 
 // desiredRouterCertsGlobalSecret returns the desired router-certs global
 // secret.
-func desiredRouterCertsGlobalSecret(secrets []corev1.Secret, ingresses []operatorv1.IngressController, operandNamespace string) (*corev1.Secret, error) {
-	if len(ingresses) == 0 || len(secrets) == 0 {
-		return nil, nil
-	}
-
-	nameToSecret := map[string]*corev1.Secret{}
-	for i, certSecret := range secrets {
-		nameToSecret[certSecret.Name] = &secrets[i]
-	}
+func desiredRouterCertsGlobalSecret(secrets []corev1.Secret, ingresses []operatorv1.IngressController, operandNamespace, clusterIngressDomain string) (*corev1.Secret, error) {
+	for i := range ingresses {
+		// The authentication operator only requires the certificate and
+		// key for the ingresscontroller that handles the cluster
+		// ingress domain, and publishing certificates for all
+		// ingresscontrollers could cause the secret's size to exceed
+		// the maximum secret size of 1 mebibyte.  See
+		// <https://issues.redhat.com/browse/OCPBUGS-853>.
+		if ingresses[i].Status.Domain != clusterIngressDomain {
+			continue
+		}
 
-	ingressToSecret := map[*operatorv1.IngressController]*corev1.Secret{}
-	for i, ingress := range ingresses {
-		// Check if ingress.Spec.DefaultCertificate is an available secret
-		// in the secrets slice. If it is not, attempt to fall back to the
-		// operator generated default certificate. If ingress.Spec.DefaultCertificate
-		// is updated to point to a non-existent secret, the certificate controller
-		// will not delete the operator generated default certificate.
-		// See https://bugzilla.redhat.com/show_bug.cgi?id=1887441
-		if defaultCert := ingress.Spec.DefaultCertificate; defaultCert != nil {
-			if secret, ok := nameToSecret[defaultCert.Name]; ok {
-				ingressToSecret[&ingresses[i]] = secret
-				continue
+		var (
+			defaultCertName = controller.RouterOperatorGeneratedDefaultCertificateSecretName(&ingresses[i], operandNamespace)
+			customCertName  = ingresses[i].Spec.DefaultCertificate
+			cert            *corev1.Secret
+		)
+		// If the ingresscontroller does not specify a secret or
+		// specifies a secret that doesn't exist, use the
+		// operator-generated default certificate.  If
+		// ingress.Spec.DefaultCertificate is updated to point to a
+		// non-existent secret, the certificate controller does not
+		// delete the operator-generated default certificate, so it is
+		// safe to fall back to using it.  See
+		// <https://bugzilla.redhat.com/show_bug.cgi?id=1887441>.
+		for i := range secrets {
+			// defaultCertName can be the same as customCertName, so
+			// both of the following conditions can be true; see
+			// <https://bugzilla.redhat.com/show_bug.cgi?id=1912922>.
+			if defaultCertName.Name == secrets[i].Name {
```

Comment:
I've stared at this for a while and still it's not really making sense. We didn't fix https://bugzilla.redhat.com/show_bug.cgi?id=1912922 here right? How'd did the old code handle this?

Reply:
Right, https://github.com/openshift/cluster-ingress-operator/pull/521/commits/00f17ef70bff89c1f9ad13e2c8c36968c3915be1 fixed https://bugzilla.redhat.com/show_bug.cgi?id=1912922 but didn't require changes in this code to do so.  In this code, the old logic was fine, but it is a little subtle, so I figured it would be helpful to add a comment to explain why I wrote the logic this way.  

The old logic went something like this:

1. Build `nameToSecret`.  
2. For each ingresscontroller, 
    a. If `ingress.Spec.DefaultCertificate` is set, add the secret named `ingress.Spec.DefaultCertificate` to `globalSecret`.  
    b. Otherwise add the secret named `controller.RouterOperatorGeneratedDefaultCertificateSecretName(&ingress, operandNamespace)` to `globalSecret`.  
3. Publish `globalSecret`.  

The new logic, because it only cares about the default ingresscontroller, inverts this logic:

1. For each secret, 
    a. If the secret's name matches `controller.RouterOperatorGeneratedDefaultCertificateSecretName(&ingress[i], operandNamespace)` where `&ingresses[i]` is the default ingresscontroller, set `cert` to the secret.
    b. If the secret's name matches `ingresses[i].Spec.DefaultCertificate` where `&ingresses[i]` is the default ingresscontroller, set `cert` to the secret`.
2. If `cert` is set, publish it in `globalSecret`.

Now suppose `controller.RouterOperatorGeneratedDefaultCertificateSecretName(&ingresses[i], operandNamespace)` is "router-certs-default" and `ingresses[i].Spec.DefaultCertificate` is "foo" and that both the "router-certs-default" secret and the "foo" secret exist.  Then the following logic could do the wrong thing:
```go
		for i := range secrets {
			if defaultCertName.Name == secrets[i].Name {
				cert = &secrets[i]
			}
			if customCertName != nil && customCertName.Name == secrets[i].Name {
				cert = &secrets[i]
			}
		}
```
The following test case (added to `TestDesiredRouterCertsGlobalSecret`) fails with the faulty logic:
```go
		{
			description: "custom certificate, with secrets having the custom one and then default one",
			inputs: testInputs{
				[]operatorv1.IngressController{defaultCICustomDefaultCert},
				[]corev1.Secret{customDefaultCert, defaultCert},
			},
			output: testOutputs{
				&corev1.Secret{
					Data: map[string][]byte{"apps.my.devcluster.openshift.com": customDefaultCertData},
				},
			},
		},
```
If `secrets` has "router-certs-default" after "foo", then `cert` would get assigned the wrong value.  

Having worked through this example, I'll add the aforementioned test case to the PR.  Does my explanation make the logic clearer?

Reply:
Yea thanks this helps - it was definitely subtle in the existing code. I think your code is a bit more verbose, which is good.

Test case makes sense.

Reply:
I went ahead and refactored the tricky logic into a new function, `getDefaultCertificateSecretForIngressController`, with more comments.

Reply:
The new variables names makes it MUCH more readable in my opinion, thanks!